flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/RecoveredChannelStateHandler.java

```diff
@@ -190,51 +195,60 @@ public void recover(
             int oldSubtaskIndex,
             BufferWithContext<BufferBuilder> bufferWithContext)
             throws IOException {
-        try (BufferBuilder bufferBuilder = bufferWithContext.context) {
-            try (BufferConsumer bufferConsumer =
-                    bufferBuilder.createBufferConsumerFromBeginning()) {
-                bufferBuilder.finish();
-                if (bufferConsumer.isDataAvailable()) {
-                    final List<CheckpointedResultSubpartition> channels =
-                            getMappedChannels(subpartitionInfo);
-                    for (final CheckpointedResultSubpartition channel : channels) {
-                        // channel selector is created from the downstream's point of view: the
-                        // subtask of downstream = subpartition index of recovered buffer
-                        final SubtaskConnectionDescriptor channelSelector =
-                                new SubtaskConnectionDescriptor(
-                                        subpartitionInfo.getSubPartitionIdx(), oldSubtaskIndex);
-                        channel.addRecovered(
-                                EventSerializer.toBufferConsumer(channelSelector, false));
-                        channel.addRecovered(bufferConsumer.copy());
-                    }
-                }
+        try (BufferBuilder bufferBuilder = bufferWithContext.context;
+                BufferConsumer bufferConsumer = bufferBuilder.createBufferConsumerFromBeginning()) {
+            bufferBuilder.finish();
+            if (!bufferConsumer.isDataAvailable()) {
+                return;
+            }
+            final List<ResultSubpartitionInfo> mappedSubpartitions =
+                    getMappedSubpartitions(subpartitionInfo);
+            for (final ResultSubpartitionInfo mappedSubpartition : mappedSubpartitions) {
+                // channel selector is created from the downstream's point of view: the
+                // subtask of downstream = subpartition index of recovered buffer
+                final SubtaskConnectionDescriptor channelSelector =
+                        new SubtaskConnectionDescriptor(
+                                subpartitionInfo.getSubPartitionIdx(), oldSubtaskIndex);
+                CheckpointedResultPartition checkpointedResultPartition =
+                        getCheckpointedResultPartition(mappedSubpartition.getPartitionIdx());
+                checkpointedResultPartition.addRecovered(
```

Comment:
Do you mean we should call `getCheckpointedResultPartition` based on the `subpartitionInfo` instead of mappedSubpartition?

If yes, sounds make sense. And updated at the last commit.

Reply:
good. It's better now